src/ui.h

```diff
@@ -146,6 +146,20 @@ typedef struct
 #define UI_MENU_SUBMENU_SUFFIX(retval, item, suffix)           { UI_ITEM_SUBMENU, retval, NULL, item, suffix }
 #define UI_MENU_SUBMENU_ACCEL(retval, item, keystroke)         { UI_ITEM_SUBMENU, retval, NULL, item, UI_MENU_ACCEL(keystroke) }
 #define UI_MENU_END                                            { UI_ITEM_END, 0, NULL, NULL, NULL }
+#define UI_MAKE_MENU_ACTION(a, f, r, i) {	\
+    a.flags = f;				\
+    a.retval = r;				\
+    a.prefix = NULL;				\
+    a.item = i;					\
+    a.suffix = NULL;				\
+  }
```

Comment:
Are you sure it is not enough to simply use the above macros and do: `my_item = UI_MENU_ACTION(ret, item);` The same for UI_MAKE_END.

Reply:
Yeah, I am.  Composite literals like you describe were introduced in ISO C99; the default behavior in the Makefile is to use -ansi, which is equivalent to -std=c90.   I've built and tested with -std=c18, the latest ISO standard, and  with -std=gnu18, which enables the GCC extensions to C18.  

I can verify it works fine on Linux, but I cannot test on other platforms. 

My view is that changing the language standard is probably safe but beyond what I can verify and that it's best to either use the implementation here or just do the assignment to members of UI_tMenuItem directly.  

Let me know how you want me to proceed. I'll make the other changes as suggested.

Reply:
Ah, I forgot to mention that if you do move to -std=c18, the same composite literal will work for both structure assignment and initialization.

Reply:
I suspected that was the reason. I don't know if supporting old C90 is really needed, but perhaps one of the older ports (using old microsoft C compilers) do not understand composite literals.  @mikrosk what do yo think?

Reply:
One thing to consider, albeit probably minor.  If you were to move to composite literals for the menu initializers, you'd have drop the static & const keywords from all declarations of UI_tMenuItem (70-something instances).

Reply:
My idea is somewhat simpler, just in this place do:
```
    for (i = 1, menu_entry = 0; i <= CARTRIDGE_LAST_SUPPORTED; i++) {
        if (CARTRIDGES[i].kb == k) {
            menu_array[menu_entry] = (UI_tMenuItem)UI_MENU_ACTION(i, CARTRIDGES[i].description);
            menu_entry++;
        }
    }
    menu_array[menu_entry] = (UI_tMenuItem)UI_MENU_END;
```
I tried a few other options, sadly all produced the some warning in C90 mode.
IMHO, this is the less ugly one:
```
	UI_tMenuItem menu_array[CARTRIDGE_LAST_SUPPORTED+1];
	UI_tMenuItem menu_end = UI_MENU_END;
	int i, menu_entry;
	int option = 0;

	UI_driver->fInit();

	for (i = 1, menu_entry = 0; i <= CARTRIDGE_LAST_SUPPORTED; i++) {
		if (CARTRIDGES[i].kb == k) {
			UI_tMenuItem menu_act = UI_MENU_ACTION(0, NULL);
			menu_act.retval = i;
			menu_act.item = CARTRIDGES[i].description;
			menu_array[menu_entry] = menu_act;
			menu_entry++;
		}
	}
	/* Terminate menu_array */
	menu_array[menu_entry] = menu_end;
```

Reply:
Yeah, I'm thinking along the same lines as your second example (although I'd move the menu_act initialization outside the loop).  It's more verbose but much easier to understand.